src/modm/architecture/interface/can_message.hpp.in

```diff
@@ -76,42 +100,82 @@ struct Message
 		return (flags.rtr != 0);
 	}
 
+	inline void
+	setDataLengthCode(uint8_t inDlc)
+	{
+		while (dlcConversionTable[inDlc] > capacity) inDlc--;
+
+		dlc = inDlc;
+		length = dlcConversionTable[inDlc];
+		if (dlc > 8) setFlexibleData();
+	}
+
+	inline void
+	setLength(uint8_t inLength)
+	{
+		uint8_t inDlc = 0;
+		while (dlcConversionTable[inDlc] < std::min<uint8_t>(inLength, 64)) inDlc++;
+		setDataLengthCode(inDlc);
+	}
+
 	inline uint8_t
 	getLength() const
 	{
 		return length;
 	}
 
-	inline void
-	setLength(uint8_t len)
+	inline uint8_t
+	getDataLengthCode() const
 	{
-		length = len;
+		return dlc;
 	}
 
+public:
+	static constexpr uint8_t capacity = {{ options["message.buffer"] }};
+	static constexpr uint8_t dlcConversionTable[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64};
 public:
 	uint32_t identifier;
-	uint8_t modm_aligned(4) data[8];
 	struct Flags
 	{
 		Flags() :
-			rtr(0), extended(1)
+			fd(0), brs(0), rtr(0), extended(1)
 		{
 		}
 
+		bool fd : 1;
+		bool brs : 1;
 		bool rtr : 1;
 		bool extended : 1;
 	} flags;
+	uint8_t dlc;
 	uint8_t length;
```

Comment:
Conventional C++ wisdom says these shouldn't be public. The class has to maintain an invariant on `dlc` and `length`. If one is changed, the other one has to be updated and not all values are valid. That is exactly what the `setLength()` and `setDataLengthCode()` member functions do.

Changing them to be private would be a breaking change, though. I am not really sure what the best option is. We have to make a choice between compatibility and sanity. What do you think @salkinium @rleh ?

Reply:
Yes, conventional C++ wisdom says these shouldn't be public. I have chosen to make both public because length was previously public. I personally, would prefer to make both private although this would interfere with backward compatibility.

Reply:
Here @rleh mentions bacward compatability as an issue here:
[https://github.com/modm-io/modm/pull/882#issuecomment-1181084541](url)

Reply:
We will not change this in this PR and leave the values as `public` for now.